### DISCUSSION TITLE
Ability to perform data refresh without restarting server

### DISCUSSION THREAD

**Discussion by ashur** [11ty/eleventy#4131] [Enhancement Queue]
## Summary

Gatsby has a handy development feature for [refreshing data](https://www.gatsbyjs.com/docs/refreshing-content/) without having to stop and restart the local server.

> To facilitate this, Gatsby exposes an environment variable called `ENABLE_GATSBY_REFRESH_ENDPOINT`.
>
> If set to `true`, this will expose a `/__refresh` webhook that can receive POST requests to refresh the sourced content. This exposed webhook can be triggered whenever remote data changes.

Paired with a [simple bookmarklet](https://gist.github.com/ashur/d9da068e091e079e4cc88f7068834882) that POSTs to the endpoint, the feature allows developers to re-fetch data with a single click without leaving the browser — ex., to test content changes in a remote CMS.

## Feature Request

It would be really helpful for clearing locally cached data if Eleventy added a hook for doing something similar.

For example, Eleventy configuration could accept a new [event](https://www.11ty.dev/docs/events/) type that allows developers to define functions for refreshing data:

```javascript
eleventyConfig.on('eleventy.refreshData', async () => {
  // Define what should happen to trigger data refresh
  // (ex., delete a folder on disk)
});
```

Ideally, these would only called if the feature is enabled (ex., by an environment variable as in Gatsby). After the refresh function has finished successfully, Eleventy would perform another build automatically, triggering a browser reload if applicable.

## Alternatives

Some possible approaches that don't require native support from Eleventy:

- Spin up a second small server (ex., Express) alongside `eleventy --serve` that listens for data refresh requests; this would primarily be filesystem and process/command-line driven
- Implement this with [Eleventy Dev Server](https://www.11ty.dev/docs/dev-server/), using [Vite middleware](https://github.com/11ty/eleventy-plugin-vite) as a model

**Comment by Snapstromegon** [comment]
While I strongly agree that this functionality would be useful, I don't see the benefit of exposing such an event (@ashur if you have a usecase that can only be solved with such an event, could you maybe explain it in a little more detail?).

I would personally prefer a solution where it is possible for plugins to register routes and handlers with EleventyServe (my first thought was towards serverless functions) and
to provide plugins also with a way to retrigger builds. That way you could start with something simpel like "rebuild every X seconds" and build as complex logic as you want.

That way you can have one API change and provide for way more usecases. Some other interesting ones that come to mind are:
- build metrics (like times and number of rebuilds)
- rebuild via keyboard shortcuts in watch mode
- rebuild for normally unwatched files with custom watch plugin
- ...

For this solution we'd need to
- rework how `Eleventy.restart` works to allow rebuilds from other sources than watch
- add a way for a plugin to trigger a `Eleventy.restart`
- add a way for a plugin to register one or more routes on an `EleventyServe` instance

**Considerations**

Doing this will take more work than what is proposed by @ashur, but will most likely result in a more flexible API covering this and many more usecases.
This will also move functions currently only possible for Eleventy into the realm of plugins at the expense of expanding the API.
